lib-ee/emqx_ee_connector/src/emqx_ee_connector_sqlserver.erl

```diff
@@ -0,0 +1,526 @@
+    | [updated_tuple()]
+    | [selected_tuple()]
+    | {error, common_reason()}.
+execute(Conn, SQL, Timeout) ->
+    odbc:sql_query(Conn, str(SQL), Timeout).
+
+to_bin(List) when is_list(List) ->
+    unicode:characters_to_binary(List, utf8).
+
+%% for bridge data to sql server
+parse_sql_template(Config) ->
+    RawSQLTemplates =
+        case maps:get(sql, Config, undefined) of
+            undefined -> #{};
+            <<>> -> #{};
+            SQLTemplate -> #{?ACTION_SEND_MESSAGE => SQLTemplate}
+        end,
+
+    BatchInsertTks = #{},
+    parse_sql_template(maps:to_list(RawSQLTemplates), BatchInsertTks).
+
+parse_sql_template([{Key, H} | T], BatchInsertTks) ->
+    case emqx_plugin_libs_rule:detect_sql_type(H) of
+        {ok, select} ->
+            parse_sql_template(T, BatchInsertTks);
+        {ok, insert} ->
+            case emqx_plugin_libs_rule:split_insert_sql(H) of
+                {ok, {InsertSQL, Params}} ->
+                    parse_sql_template(
+                        T,
+                        BatchInsertTks#{
+                            Key =>
+                                #{
+                                    ?BATCH_INSERT_PART => InsertSQL,
+                                    ?BATCH_PARAMS_TOKENS => emqx_plugin_libs_rule:preproc_tmpl(
+                                        Params
+                                    )
+                                }
+                        }
+                    );
+                {error, Reason} ->
+                    ?SLOG(error, #{msg => "split sql failed", sql => H, reason => Reason}),
+                    parse_sql_template(T, BatchInsertTks)
+            end;
+        {error, Reason} ->
+            ?SLOG(error, #{msg => "detect sql type failed", sql => H, reason => Reason}),
+            parse_sql_template(T, BatchInsertTks)
+    end;
+parse_sql_template([], BatchInsertTks) ->
+    #{
+        ?BATCH_INSERT_TEMP => BatchInsertTks
+    }.
+
+%% single insert
+apply_template(
+    {?ACTION_SEND_MESSAGE = _Key, _Msg} = Query, Templates
+) ->
+    %% TODO: fix emqx_plugin_libs_rule:proc_tmpl/2
+    %% it won't add single quotes for string
+    apply_template([Query], Templates);
+%% batch inserts
+apply_template(
+    [{?ACTION_SEND_MESSAGE = Key, _Msg} | _T] = BatchReqs,
+    #{?BATCH_INSERT_TEMP := BatchInsertsTks} = _Templates
+) ->
+    case maps:get(Key, BatchInsertsTks, undefined) of
+        undefined ->
+            BatchReqs;
+        #{?BATCH_INSERT_PART := BatchInserts, ?BATCH_PARAMS_TOKENS := BatchParamsTks} ->
+            SQL = emqx_plugin_libs_rule:proc_batch_sql(BatchReqs, BatchInserts, BatchParamsTks),
+            {Key, SQL}
+    end;
+apply_template(Query, Templates) ->
+    %% TODO: more detail infomatoin
+    ?SLOG(error, #{msg => "apply sql template failed", query => Query, templates => Templates}),
+    {error, failed_to_apply_sql_template}.
+
+do_async_reply(Result, {ReplyFun, [Context]}) ->
+    ReplyFun(Context, Result).
```

Comment:
This `ReplyFun` might be the one passed down by `emqx_resource_buffer_worker`, right?
If so, I think it's safer to avoid assuming the argument list is a single-element one:

```suggestion
do_async_reply(Result, {ReplyFun, Args}) ->
    apply(ReplyFun, Args ++ Result).
```

Reply:
I understand it is **now** a single element list.  But that's an internal implementation detail that should not be relied on by modules called by buffer workers.  :thinking: